akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -996,7 +996,14 @@ object GraphDSL extends GraphApply {
      * INTERNAL API
      */
     private[GraphDSL] def addEdge[T, U >: T](from: Outlet[T], to: Inlet[U]): Unit =
-      traversalBuilderInProgress = traversalBuilderInProgress.wire(from, to)
+      if (traversalBuilderInProgress.isUnwired(from) && traversalBuilderInProgress.isUnwired(to)) {
+        traversalBuilderInProgress = traversalBuilderInProgress.wire(from, to)
+      } else {
+        if (!traversalBuilderInProgress.isUnwired(from))
+          throw new IllegalArgumentException(s"[${from.s}] is already connected")
+        else if (!traversalBuilderInProgress.isUnwired(to))
+          throw new IllegalArgumentException(s"[${to.s}] is already connected")
+      }
```

Comment:
would it be worth doing the `wire` optimistically and only in catch exception do the isUnwired checks? some of them are Map lookups.